zerocopy-derive/src/lib.rs

```diff
@@ -549,6 +575,28 @@ fn impl_block<D: DataExt>(
         .chain(field_type_bounds.iter())
         .chain(padding_check_bound.iter());
 
+    let layout_extras = if trt == Trait::KnownLayout {
+        // We currently only support deriving for sized types; this code will
+        // fail to compile for unsized types.
+        Some(quote!(
+            const LAYOUT: zerocopy::DstLayout = zerocopy::DstLayout::for_type::<Self>();
+
+            // SAFETY: `.cast` preserves address and provenance.
+            //
+            // TODO(#429): Add documentation to `.cast` that promises that
+            // it preserves provenance.
+            #[inline(always)]
+            fn raw_from_ptr_len(
+                bytes: ::core::ptr::NonNull<u8>,
+                _elems: usize,
+            ) -> ::core::ptr::NonNull<Self> {
+                bytes.cast::<Self>()
+            }
```

Comment:
@joshlf This is the only change to production code, and is lifted verbatim from the declarative macro.

Reply:
Ack, lgtm.